### PR TITLE
Add Header and submit votes in SMC 

### LINKED ISSUE #89: SMC Changes for a minimal sharding protocol
The research team has identified a few new changes in "`spec 1.1`" that they are confident will not be removed in a future specification. 

High level requirements (please check off these as PRs are merged):

- [x] Anyone can call addHeader(period_id, shard_id, chunks_root) at any time. The first header to get included for a given shard in a given period gets in, all others don’t. This function just emits a log.

- [x] For every combination of shard and period, N collators (now called “notaries”) are sampled. They try to download the collation body corresponding to any header that was submitted. They can call a function submitVote(period_id, shard_id, chunks_root). This function just emits a log.

- [x] Clients read logs. If a client sees that in some shard, for some period, a chunk has been included and >= 2N/3 notaries voted for it, it accepts it as part of the canonical chain.

Assigning to SMC expert @enriquefynn for now. 
We should consider writing smaller tasks to achieve the above requirements unless this can be completed in 1 or 2 medium sized pull requests. 

References:

- [Sharding spec 1](https://ethresear.ch/t/sharding-phase-1-spec/1407) <- Now out of date!
- [Minimal sharding procotol](https://ethresear.ch/t/a-minimal-sharding-protocol-that-may-be-worthwhile-as-a-development-target-now/1650) <- The post that prompted this issue
- https://github.com/ethereum/sharding/issues/76 <- Excellent issue description by @hwwhww

### PR DESCRIPTION
**Pull request by terenc3t**
To further along with the minimal sharding protocol spec, This PR will address both `add header` and `submit_vote` functionalities in SMC. `add_header` will be called by proposer to submit collation header to the main chain. In the process of `add_header` we'll update the notary sample size and the collation records. Sampled notaries will call `submit_vote` to vote on collation header of a given shard and period. SMC will keep track of the total votes and flag collation header as elected when it reaches quorum size.

- [x] Function in SMC for proposer to add header
- [x] Function in SMC for notary to submit vote
- [x] Test cases for `add_header`
- [x] Test cases for `submit_vote`

Closes #89

### CODE REVIEW

File: sharding/contracts/sharding_manager.sol [hunk 1]
```diff
@@ -2,56 +2,57 @@ pragma solidity ^0.4.23;
...
-
     // Notary sample size at current period and next period
     uint currentPeriodNotarySampleSize;
     uint nextPeriodNotarySampleSize;
     uint sampleSizeLastUpdatedPeriod;
 
     // Constant values
+    uint constant ONE = 1;
+    uint constant ONES = 255;
+    uint constant LAST_BYTE = 31;
```
[the remark this answers is not in the document]
Comment (terenc3t) [reply sharding/contracts/sharding_manager.sol]: Constant variables for bit wise operations. Set index bit to 1.
a.) 1 << (255 - index)
b.) ONE << (ONES - index)
Which one is better?

[the remark this answers is not in the document]
Comment (prestonvanloon) [reply sharding/contracts/sharding_manager.sol]: Can you please add a comment? The variable name isn't very descriptive which lead me to ask "what is this". 

I still don't know what it is.

File: sharding/contracts/sharding_manager.sol [hunk 2]
```diff
@@ -221,4 +238,22 @@ contract SMC {
         --emptySlotsStackTop;
         return emptySlotsStack[emptySlotsStackTop];
     }
-}
\ No newline at end of file
+
+    /// Check if a bit is set, this function is used to check
+    /// if a notary has casted the vote, return true if voted, false if not
+    function hasVoted(uint _shardId, uint _index) internal returns (bool) {
+        return currentVote[_shardId] & bytes32(ONE << uint(ONES - _index)) == bytes32(ONE << uint(ONES - _index));
```
[the remark this answers is not in the document]
Comment (terenc3t) [reply sharding/contracts/sharding_manager.sol]: It's this:
**currentVote[_shardId] & bytes32(ONE << uint(ONES - _index))** == **bytes32(ONE << uint(ONES - _index))**
I should be able to shorten it

File: sharding/contracts/sharding_manager.sol [hunk 3]
```diff
@@ -221,4 +238,22 @@ contract SMC {
...
+    }
+
+    /// Set a bit to one, notary uses this function to cast it's vote,
+    /// After the notary casts it's vote, we increase the vote count by 1 (last byte)
+    /// Returns the total numbers of votes
+    function castVote(uint _shardId, uint _index) internal returns (uint8) {
```
Comment (prestonvanloon) [inline sharding/contracts/sharding_manager.sol]: Can you add more to the description of voting?

How is this supposed to work? Are there only 31 notaries and they place their vote on their index?

File: sharding/contracts/sharding_manager.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply sharding/contracts/sharding_manager.sol]: Yeah, but if B is false and A is true then this still evaluates to true?

true & false == false  is true?

File: sharding/contracts/sharding_manager.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (terenc3t) [reply sharding/contracts/sharding_manager.sol]: I'll add comments

File: sharding/contracts/sharding_manager.sol (hunk 3 above)
Comment (terenc3t) [reply sharding/contracts/sharding_manager.sol]: I'll add more color to the comments. There are 31 * 8 notaries. Each place it's vote on the unique index.

File: sharding/contracts/sharding_manager.sol [hunk 4]
```diff
@@ -162,31 +160,50 @@ contract SMC {
...
-      */
-    }
-
     /// Add collation header to the main chain, anyone can call this function. It emits a log
     function addHeader(
         uint _shardId,
-        uint period,
-        bytes32 chunkRoot,
-        address proposerAddress
+        uint _period,
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply sharding/contracts/sharding_manager.sol]: No, I was just wondering. Personally, I hate it, but this seems to be the standard for solidity so lets keep it :)

File: sharding/contracts/sharding_manager.sol (hunk 3 above)
Comment (prestonvanloon) [reply sharding/contracts/sharding_manager.sol]: Ah, yes that would be helpful context

File: sharding/contracts/sharding_manager.sol [hunk 5]
```diff
@@ -162,31 +160,50 @@ contract SMC {
...
+        uint _shardId,
+        uint _period,
+        uint _index,
+        bytes32 _chunkRoot        
+    ) public {
+        require(notaryRegistry[msg.sender].deposited);
+        require(collationRecords[_shardId][_period].chunkRoot == _chunkRoot);
+        require(!hasVoted(_shardId, _index));
+        require(getNotaryInCommittee(_shardId, _index) == msg.sender);
+
```
Comment (NIC619) [inline sharding/contracts/sharding_manager.sol]: It seems that we need to check `_period` against current period and `lastSubmittedCollation[_shardId]`.

Comment (NIC619) [reply sharding/contracts/sharding_manager.sol]: Also check against `_index` against `COMMITTEE_SIZE`.

File: sharding/contracts/sharding_manager.sol [hunk 6]
```diff
@@ -69,9 +70,6 @@ contract SMC {
     // Threshold(number of notaries in committee) for a proposal to be accepted
     uint constant QUORUM_SIZE = 90;
 
-    // Log the latest period number of the shard
-    mapping (int => int) public periodHead;
-
     /// Checks if a notary with given shard id and period has been chosen as
     /// a committee member to vote for header added on to the main chain
     function getNotaryInCommittee(uint shardId, uint _index) public view returns(address) {
```
Comment (NIC619) [inline sharding/contracts/sharding_manager.sol]: We need to check `shardId` against `SHARD_COUNT` either in here or in `submit_vote`(or both).

Comment (terenc3t) [reply sharding/contracts/sharding_manager.sol]: Good call. Added check for `shardId` in `submit_vote`

File: sharding/contracts/sharding_manager.sol (hunk 5 above)
Comment (terenc3t) [reply sharding/contracts/sharding_manager.sol]: Thanks. Checks added

File: sharding/contracts/sharding_manager.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (terenc3t) [reply sharding/contracts/sharding_manager.sol]: Fixed. Right shifting `currentVote` by `index` then AND with `1`, return true if voted, false if not

File: sharding/contracts/sharding_manager_test.go [hunk 7]
```diff
@@ -441,41 +450,37 @@ func TestCommitteeListsAreDifferent(t *testing.T) {
...
-	// register 10000 notaries to SMC.
+	// Register 1000 notaries to SMC.
 	for i := 0; i < notaryCount; i++ {
 		smc.RegisterNotary(txOpts[i])
 		backend.Commit()
 	}
 
 	numNotaries, _ := smc.NotaryPoolLength(&bind.CallOpts{})
-	if numNotaries.Cmp(big.NewInt(10000)) != 0 {
+	if numNotaries.Cmp(big.NewInt(1000)) != 0 {
```
Comment (prestonvanloon) [inline sharding/contracts/sharding_manager_test.go]: Reuse `notaryCount`

**Comment by rauljordan** [comment]
Great job @terenc3t, thank you

**Pull request merged**